Policy development tutorial (#1204)
3scale/APIcast: doc/policy_tutorial/DEV_ENV_SETUP.md

# Setting up the development environment

> As was clear from the introduction, APIcast policies are created in the Lua programming language. So we need to setup an environment to do some Lua programming. Also, an actual APIcast server would be very nice to perform some local tests.

Reviewer: ```suggestion
You need to configure an environment to work with Lua because APIcast policies are created in the Lua programming language. Additionally, you can use an actual APIcast server to perform local tests.
```

> Luckily APIcast makes it very easy to setup a development environment for APIcast using Docker and Docker Compose.

Reviewer: ```suggestion
Following simple steps, you can set up a development environment for APIcast using Docker and Docker Compose.
```

### prerequisites
This means both Docker and Docker compose must be installed.

> The version of Docker I currently use is:

Reviewer: You can remove this line :point_up:

Reviewer 2: addressed in https://github.com/3scale/APIcast/pull/1384

>     Docker version 19.03.8

Reviewer: ```suggestion
   * Docker version 19.03.8
```

> Instructions for installing Docker can be found on the Docker [website](https://docs.docker.com/get-docker/).

Reviewer: You can remove this line point_up

Reviewer 2: addressed in #1384

> With Docker compose version:

Reviewer: You can remove this line point_up

Reviewer 2: addressed in #1384

>     docker-compose version 1.25.4

Reviewer: ```suggestion
   * `docker-compose` version 1.25.4. You can find instructions for installing `docker-compose` in the Docker [website](https://docs.docker.com/compose/install/).
```

Instructions for installing Docker-compose can also be found on the Docker [website](https://docs.docker.com/compose/install/).

### Setting up the development image

> Now that we have both Docker and Docker-compose installed we an setup the APIcast development image.

Reviewer: ```suggestion
After you have installed both Docker and `docker-compose`, you can configure the APIcast development image.
```

> Firstly the APIcast git repostitory must be cloned so we can start the development of our policy. Since we are going to base our policy on the latest 3scale release we are switching to the stable branch of APIcast.

Reviewer: ```suggestion
. Clone the APIcast git repository. Use the stable branch of APIcast to base the polic on the latest 3scale release.
```

```shell
git clone https://github.com/3scale/apicast.git
```

> when done switch to a stable branch, I am using 3.8

Reviewer: You can remove this line.

```shell
cd apicast/
git checkout 3.8-stable
```

> To start the APIcast containers using Docker-compose we can use the Make file provided by 3scale. In the APIcast directory simply execute the command:

Reviewer: ```suggestion
. To start the APIcast containers using `docker-compose`,  use the *make* file provided by 3scale. Run this command in the APIcast directory:
```

```shell
make development
```

> ![make-development](img/make-development.png)

Reviewer: Is this image essential to provide a clear understanding of the configuration process? 

- If `yes`, please include a paragraph explaining the image.
- If `no`, you can remove the image.

Reviewer 2: leaving the image. It gives a clear understanding of what it does. For those familiar with docker and docker-compose (which is a requisite for any APIcast contributor)

I would not add an image, instead I would copy those lines in a code block. But good enough

Reviewer 2: replace the image with the text

> The Docker container starts in the foreground with a bash session. The first thing we need to do inside the container is installing all the dependencies.

Reviewer: ```suggestion
. The Docker container starts in the foreground with a bash session. The next step is to install all the dependencies inside the container, using a *make* command:
```

> This can also be done using a Make command, which again must be issued **inside** the container.

Reviewer: You can remove this line.

Reviewer 2: addressed in #1384

```shell
make dependencies
```

> It will now download and install a plethora of dependencies inside the container.

Reviewer: You can remove this line

Reviewer 2: addressed in #1384

> The output will be very long, but if everything went well you should be greeted with an output that looks something like this:

Reviewer: ```suggestion
+
The output will be very long. After a successful completion of the installation of all the dependencies, you will see a message similar to this:
+
```

> ![make-dependencies](img/make-dependencies.png)

Reviewer: Can you replace the image with the text that the user will see?

Reviewer 2: done

> Now as a final verification we can run some APIcast unit tests to see if we are up and running and ready to start the development of our policy.

Reviewer: Does "as a final verification" refer to a confirmation step related to the installation of the dependencies inside the container? If 'yes`, please reword this sentence to remove the mentioned phrase, because the general policy configuration is still in progress.

Reviewer 2: addressed in #1384

> To run the Lua unit tests run the following command **inside** the container:

Reviewer: ```suggestion
. To run the Lua unit tests use this command **inside** the container:
```

```shell
make busted
```
![make-busted](img/make-busted.png)

> Now that we can successfully run unit tests we can start our policy development!

Reviewer: ```suggestion
After confirming that the unit tests run successfully, you can start the policy development.
```
